### ISSUE TITLE
Positioning Caret in CTL Text

### ISSUE THREAD

**Issue by jcuenod** [atom/atom#9397]
Use this sample Hebrew (rtl)

```
כָּתֵף
```

Starting from the extreme left, I would expect a right arrow either to move the cursor to the next visual position (that is, between the ת and the ף) or jump to what is the next logical position (between the כ and the ת). Instead Atom doesn't move the cursor at all. Functionally however, it has moved between the כ and the dagesh (middle dot), similarly if I hit the right arrow again, the cursor stays at point 0 but logically it is now between the qamets (the vowel point beneath the כ) and the dagesh.

This gets interesting when you hit the right arrow a third time. The cursor now moves between the ת and the ף but logically it is actually between the כ and the ת and if you hit backspace, it will remove the qamets).

This issue reported from https://github.com/atom/atom/issues/4007

**Comment by lee-dohm** [comment]
@as-cii is this a duplicate? Should the old Issue be reactivated?

**Comment by as-cii** [comment]
Thanks @jcuenod!

@lee-dohm: I think this isn't being caused by our rendering mechanism, so it can be seen as a different engine. What seem to happen here is that we're not correctly recognizing paired characters: I guess this is a problem with our `text-utils`, which as far as I remember is designed for left-to-right languages and it's probably not recognizing code points in the right order.

**Comment by white-gecko** [comment]
I have the somehow the same Issue, if I have a document just containing `כָּתֵף`, the cursor moves even visually between the characters (btw.: which is somehow cool to edit the punctuation :-), but this could become a special feature).
As soon, as I have a multilingual String/Line, containing LTR and RTL, the visual cursor stops at the border between LTR and RTL, but logically moves between through the punctuation and characters, finally the visual cursor appears at the end of the line again: `Schulter כָּתֵף`.

**Comment by jcuenod** [comment]
Yes, just to add to something @white-gecko is mentioning; visually moving from left to right is not a problem, it could actually be useful (I still find it jarring and a little confusing in LibreOffice to be moving between words and suddenly the cursor changes its direction of movement and jumps around a lot) and less of a roller coaster but right now it's kind of unusable (without a lot of mental gymnastics) unfortunately.

**Comment by white-gecko** [comment]
I don't know, if I get, which way of cursor movement you like, but I try to put my opinion into an image.
![image2993](https://cloud.githubusercontent.com/assets/1018168/13437168/1a5553a6-dfe3-11e5-9569-287de48307e9.png)

**Comment by lee-dohm** [comment]
For those of us who admittedly aren't as well-versed as perhaps we should be on this stuff, what is the definition of CTL? Is it the same or different from RTL?

**Comment by jcuenod** [comment]
@white-gecko this is, I think, an essential option (but I kind of like visual left and right more than logical left and right).

@lee-dohm I just had a look at [Wikipedia](https://en.wikipedia.org/wiki/Complex_text_layout) and found that it's more than this but I think of it as "Bi-directional text, where characters may be written from either right-to-left or left-to-right direction."

**Comment by jcuenod** [comment]
To my knowledge, this is not resolved - the caret does not move correctly.

**Comment by 50Wliu** [comment]
Apologies, there's been some issues with probot not hitting the rate limit and not being able to post a warning comment.

**Comment by yakirbu** [comment]
Still relevant.

**Comment by jcuenod** [comment]
1. The issue is unresolved in the latest version.
2. Version and issue:
 - Atom 1.32.2 x64
 - Antergos
 - Move the caret over CTL text: The caret jumps to the end of the RTL text (left or right, depending on the direction of movement) and stays there until the cursor is no longer positioned in the RTL text.
